### ISSUE TITLE
Local goes back when clicking on nextjs Link component

### ISSUE THREAD

**Issue by huskyjp** [QuiiBz/next-international#293]
**Describe the bug**
Maybe related to #138 & #140, when I click nextjs `<Link>` component, it forces to go back to the default locale. 
**To Reproduce**
1. Go to [CodesandBox](https://codesandbox.io/p/devbox/next-international-app-router-forked-l2f3p6?file=%2Fapp%2Fheader.tsx%3A26%2C4-28%2C14)
2. Change locale from `en` -> `fr`
3. Click `Link component` 
4. Always route to default locale `en` even we pass the locale prop in the <Link />

**Expected behavior**
Should keep the locale

**About (please complete the following information):**
- next-international: "^1.1.4",
 - Next.js version: "^14.0.3"

**Comment by QuiiBz** [comment]
Do you reproduce the issue when running locally? AFAIK this is a bug with CodeSandbox, they don't preserve cookies correctly.

**Comment by huskyjp** [comment]
@QuiiBz Thanks for the response! Yes it also does the same behavior in my locale environment. But I also noticed that this problem is related to the cookies. 

The thing is my current `middleware.ts` has some rewrite condition like below 

```
import { NextRequest } from "next/server"

import { createI18nMiddleware } from "next-international/middleware"

export const I18nMiddleware = createI18nMiddleware({
  locales: ["en", "jp"],
  defaultLocale: "en",
  urlMappingStrategy: "rewriteDefault",
})

export default function middleware(request: NextRequest) {
  const url = request.nextUrl.clone()
  const pathname = url.pathname

  // Check if the URL does not contain any of the specified locales
 // It basically works but did not change the nextjs cookie locale info so when we use `Link` component, it behaves strange.
  // if (!pathname.startsWith("/en/") && !pathname.startsWith("/jp/")) {
  //   url.pathname = `/en${pathname}`

  //   console.log("Redirecting to: ", url.toString())
  //   return NextResponse.rewrite(url)
  // }

  return I18nMiddleware(request)
}

export const config = {
  matcher: ["/((?!api|static|.*\\..*|_next|favicon.ico|robots.txt).*)"],
}
```

The reason why I added ` if (!pathname.startsWith("/en/") && !pathname.startsWith("/jp/"))` is when we try to dynamically access the path without specifying the default locale (in this case, dynamically typing & changing the URL path from `localhost:3000/jp/admin/` to `localhost:3000/admin/`), nextjs stays the locale as is since the cookie is still the same so I can't basically change the locale from the URL bar. 

So basically I want to know if there is a way to handle user can type and dynamically change the locale & route to the correct path from URL bar without above if statement in the middleware.ts?

**Comment by QuiiBz** [comment]
I'm not able to reproduce the issue using the example in the repo: https://github.com/QuiiBz/next-international/tree/main/examples/next-app

The steps I did:
- Clear cookies
- Navigate to `/`, automatically redirected to `/en`, cookie set to `en`
- Click on `fr`, redirected to `/fr`, cookie updated to `fr`
- Navigate to other pages, the route segment is still `fr`, cookie is still `fr`

Does this work on your side, and if no could share the steps to reproduce in this example above? CodeSandbox has issues with cookies so that's why I'd like to reproduce outside of it.

**Comment by gustaveWPM** [comment]
Hello @QuiiBz 

I experiment an issue which could be related to this one.

To reproduce:
- Open a new private navigation window
- Go to https://dashboard-rtm.vercel.app/
- Click on the language switcher, then switch to English/French (depending of your current locale on the website first-visit)
- Click on the "Login" button of the navbar
- Login via Discord
- The website locale is out-of-sync

Source code:
https://github.com/Tirraa/dashboard_rtm

Maybe this is due to my middlewares chain?
I think the `onClick={() => signIn('discord', { callbackUrl: ROUTES_ROOTS.DASHBOARD })}` on the Login button, is redirecting to Discord OAuth "Before" something in next-international which saves the user locale choice...
